File: src/tools/system.ts
import { type ToolDefinition, type ToolResult } from "../core/types.js";

export function getSystemTools(): ToolDefinition[] {
  const make = (n: string, d: string, s: Record<string, unknown>, fn: (p: Record<string, unknown>, ctx: any) => Promise<ToolResult>): ToolDefinition => ({
    name: n, description: d, inputSchema: s as any, execute: fn,
  });

  return [
    make("system_overview", "Get OS, CPU, memory, versions, machine ID. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { info { os { name version arch } cpu { name cores } memory { total free } machineId } }`) })),
    make("system_array", "Get array state, capacity, disk health. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { array { state capacity disks { id name size type status } } }`) })),
    make("system_network", "Get access URLs, ports, LAN/WAN IPs. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { network { accessUrls { type name ipv4 ipv6 } } }`) })),
    make("system_registration", "Get license info. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { registration { type keyFile expiration } }`) })),
    make("system_variables", "Get full Unraid variables. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { variables }`) })),
    make("system_metrics", "Get CPU and memory usage. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { metrics { cpu memory { total used free usedPercent } } }`) })),
    make("system_network_metrics", "Get network throughput. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { metrics { network { rxBytes txBytes rxSec txSec } } }`) })),
    make("system_services", "Get running services. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { services { name online version } }`) })),
    make("system_display", "Get UI theme. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { display { theme } }`) })),
    make("system_config", "Get config validity. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { config { valid validations { type message } } }`) })),
    make("system_online", "Check server reachability. Returns boolean. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { online }`) })),
    make("system_owner", "Get owner info. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { owner { name avatar profileUrl } }`) })),
    make("system_server", "Get hostname, uptime, version. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { server { hostname uptime version arrayState } }`) })),
    make("system_servers", "Get all registered servers. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { servers { id name ip lan url gui } }`) })),
    make("system_flash", "Get flash drive info. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { flash { device vendor product size } }`) })),
    make("system_ups_devices", "Get UPS devices. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { ups { devices { name batteryCharge status } } }`) })),
    make("system_server_time", "Get server time and NTP. Read-only.", { type: "object", properties: {} },
      async (_p: any, ctx: any) => ({ isSuccess: true, data: await ctx.query(`query { serverTime { time timeZone ntp } }`) })),
  ];
}

File: go/internal/tools/system.go
package system

import (
	"context"
	"fmt"

	"github.com/hektyc/unraid-mcp-server/go/internal/config"
	"github.com/hektyc/unraid-mcp-server/go/internal/guards"
	"github.com/hektyc/unraid-mcp-server/go/internal/mcp"
)

func RegisterTools(s *mcp.Server, cfg *config.Config) {
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_overview",
		Description: "Get OS, CPU, memory, versions, machine ID. Read-only.",
		Query:       `query { info { os { name version arch } cpu { name cores } memory { total free } machineId } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_array",
		Description: "Get array state, capacity, disk health. Read-only.",
		Query:       `query { array { state capacity disks { id name size type status } } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_network",
		Description: "Get access URLs, ports, LAN/WAN IPs. Read-only.",
		Query:       `query { network { accessUrls { type name ipv4 ipv6 } } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_registration",
		Description: "Get license info. Read-only.",
		Query:       `query { registration { type keyFile expiration } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_variables",
		Description: "Get full Unraid variables. Read-only.",
		Query:       `query { variables }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_metrics",
		Description: "Get CPU and memory usage. Read-only.",
		Query:       `query { metrics { cpu memory { total used free usedPercent } } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_network_metrics",
		Description: "Get network throughput. Read-only.",
		Query:       `query { metrics { network { rxBytes txBytes rxSec txSec } } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_services",
		Description: "Get running services. Read-only.",
		Query:       `query { services { name online version } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_display",
		Description: "Get UI theme. Read-only.",
		Query:       `query { display { theme } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_config",
		Description: "Get config validity. Read-only.",
		Query:       `query { config { valid validations { type message } } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_online",
		Description: "Check server reachability. Read-only.",
		Query:       `query { online }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_owner",
		Description: "Get owner info. Read-only.",
		Query:       `query { owner { name avatar profileUrl } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_settings",
		Description: "Get unified settings. Read-only.",
		Query:       `query { settings }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_server",
		Description: "Get hostname, uptime, version. Read-only.",
		Query:       `query { server { hostname uptime version arrayState } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_servers",
		Description: "Get all registered servers. Read-only.",
		Query:       `query { servers { id name ip lan url gui } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_flash",
		Description: "Get flash drive info. Read-only.",
		Query:       `query { flash { device vendor product size } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_ups_devices",
		Description: "Get UPS devices. Read-only.",
		Query:       `query { ups { devices { name batteryCharge status } } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_ups_config",
		Description: "Get UPS config. Read-only.",
		Query:       `query { ups { config { driver port } } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_server_time",
		Description: "Get server time and NTP. Read-only.",
		Query:       `query { serverTime { time timeZone ntp } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_timezones",
		Description: "Get available timezones. Read-only.",
		Query:       `query { timezones }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "system_network_interfaces",
		Description: "Get network interfaces. Read-only.",
		Query:       `query { network { interfaces { name type ipv4 ipv6 mac } } }`,
	})
}